phase/prepare_hosts.go

```diff
@@ -30,6 +30,28 @@ type prepare interface {
 	Prepare(os.Host) error
 }
 
+// updateEnvironment updates the environment variables on the host and reconnects to
+// it if necessary.
+func (p *PrepareHosts) updateEnvironment(h *cluster.Host) error {
+	if err := h.Configurer.UpdateEnvironment(h, h.Environment); err != nil {
+		return err
+	}
+	if h.Connection.Protocol() != "SSH" {
+		return nil
+	}
+	// XXX: this is a workaround. UpdateEnvironment on rig's os/linux.go writes
+	// the environment to /etc/environment and then exports the same variables
+	// using 'export' command. This is not enough for the environment to be
+	// preserved across multiple ssh sessions. We need to write the environment
+	// and then reopen the ssh session. Go's ssh client.Setenv() depends on ssh
+	// server configuration (sshd only accepts LC_* variables by default).
+	h.Connection.Disconnect()
+	if err := h.Connection.Connect(); err != nil {
```

Comment:
Maybe the retry logic from `phase/connect.go` should be copied here.

Something like `log.Infof("%s: Reconnecting", h)` would be nice too.

Reply:
Maybe checking if the new env variables took effect after reconnecting would be a good measure?

Reply:
Hi @kke , I have checked and the env variables are available after the reconnection (aka `apt` command to install `curl` starts to use the defined proxy). Do you feel like it is necessary do to this on the code ? 

I have copied the logic for retrying the connection. Tested and it works. Something I noticed is this on the log:

```
INFO [ssh] 10.0.0.2:22: updating environment
INFO [SSH] 10.0.0.2: Reconnecting
INFO [ssh] 10.0.0.2:22: installing package curl
```

You see how the protocol changed from "ssh" to "SSH" and the port also went missing. It worked nonetheless but I believe there is something missing here. Can you give me some directions to investigate why this changed ?

Reply:
Oh, I see. I think I need to log BEFORE the disconnection (or after the connection). Let me see.

Reply:
Yeah, that did the trick. I also change the logging to be consistent with others.

```
INFO ==> Running phase: Prepare hosts
INFO [ssh] 10.0.0.2:22: updating environment
INFO [ssh] 10.0.0.2:22: reconnecting to apply new environment
INFO ==> Running phase: Gather host facts
```

I think that after Disconnect() we kinda loose some variables that are only reset after the Connect().

Reply:
Yes, `Disconnect` sets `c.client=nil` which makes `String()` unable to use `c.client.String()`.